### PR TITLE
keep-until label

### PR DESCRIPTION
**Pull request by njuettner**
Checking `keep-until` label, see https://intranet.giantswarm.io/docs/dev-and-releng/test-environments/#label-the-cluster

## Checklist

- [x] Update changelog in CHANGELOG.md.

### CODE REVIEW

**Comment by AverageMarcus** [comment]
Should the label not be `keep-until` ?

**Comment by njuettner** [comment]
> Should the label not be `keep-until` ?

My bad 🤦🏻, I'm was somewhere else 🤣

**Comment by AverageMarcus** [comment]
Are you updating the code as well ?

File: controllers/cluster_controller_test.go [hunk 1]
```diff
@@ -108,7 +108,7 @@ func TestClusterController(t *testing.T) {
 					Name:      "test",
 					Namespace: "default",
 					CreationTimestamp: metav1.Time{
-						Time: time.Now().Add(-eventDefaultTTL),
+						Time: time.Now().Add(-9 * time.Hour),
```
Comment (njuettner) [inline controllers/cluster_controller_test.go:111]:
```suggestion
						Time: time.Now().Add(-eventDefaultTTL),
```

File: controllers/cluster_controller.go [hunk 2]
```diff
@@ -64,21 +64,42 @@ func (r *ClusterReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ct
...
+		if time.Now().UTC().After(t) {
+			keepUntilExpired = true
+		} else {
+			IgnoredTotal.WithLabelValues(cluster.Name, cluster.Namespace).Inc()
+			log.Info(fmt.Sprintf("Found label %s. Cluster %s/%s will be ignored for deletion", keepUntil, cluster.Namespace, cluster.Name))
+			return ctrl.Result{RequeueAfter: 24 * time.Hour}, nil
+		}
+	} else {
+		keepUntilExpired = true
 	}
```
Comment (AverageMarcus) [inline controllers/cluster_controller.go:92]:
```suggestion
	// check if cluster has a keep-until label with a valid ISO date string
	if v, ok := cluster.Labels[keepUntil]; ok {
		t, err := time.Parse(keepUntilTimeLayout, v)
		if err != nil {
			ErrorsTotal.WithLabelValues(cluster.Name, cluster.Namespace).Inc()
			log.Error(err, fmt.Sprintf("failed to parse keep-until label value for cluster %s/%s", cluster.Namespace, cluster.Name))
			return ctrl.Result{}, nil
		}
		if time.Now().UTC().Before(t) {
			IgnoredTotal.WithLabelValues(cluster.Name, cluster.Namespace).Inc()
			log.Info(fmt.Sprintf("Found label %s. Cluster %s/%s will be ignored for deletion", keepUntil, cluster.Namespace, cluster.Name))
			return ctrl.Result{RequeueAfter: 24 * time.Hour}, nil
		}
	} 
```

I'm not 100% sure if this logic is correct but I think it results in the same outcome as we're returning if we've not expired, right?
(would need to change line 102 also)

Comment (AverageMarcus) [inline controllers/cluster_controller.go:92]:
```suggestion
	// check if cluster has a keep-until label with a valid ISO date string
	if v, ok := cluster.Labels[keepUntil]; ok {
		t, err := time.Parse(keepUntilTimeLayout, v)
		if err != nil {
			ErrorsTotal.WithLabelValues(cluster.Name, cluster.Namespace).Inc()
			log.Error(err, fmt.Sprintf("failed to parse keep-until label value for cluster %s/%s", cluster.Namespace, cluster.Name))
			return ctrl.Result{}, nil
		}
		if time.Now().UTC().Before(t) {
			IgnoredTotal.WithLabelValues(cluster.Name, cluster.Namespace).Inc()
			log.Info(fmt.Sprintf("Found label %s. Cluster %s/%s will be ignored for deletion", keepUntil, cluster.Namespace, cluster.Name))
			return ctrl.Result{RequeueAfter: 24 * time.Hour}, nil
		}
	} 
```

I'm not 100% sure if this logic is correct but I think it results in the same outcome as we're returning if we've not expired, right?
(would need to change line 102 also)

File: controllers/cluster_controller.go [hunk 3]
```diff
@@ -64,21 +64,42 @@ func (r *ClusterReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ct
...
+	// ignore cluster deletion if timestamp is not nil or zero
 	if !cluster.DeletionTimestamp.IsZero() {
 		PendingTotal.WithLabelValues(cluster.Name, cluster.Namespace).Inc()
 		log.Info(fmt.Sprintf("Deletion for cluster %s/%s is already applied", cluster.Namespace, cluster.Name))
 		return ctrl.Result{}, nil
 	}
 
 	// immediately delete the cluster if defaultTTL has passed
-	if deletionTimeReached(cluster) {
+	if deletionTimeReached(cluster) && keepUntilExpired {
```
Comment (njuettner) [inline controllers/cluster_controller.go:102]:
```suggestion
	if deletionTimeReached(cluster) {
```

**Review by AverageMarcus — APPROVED** [review]

**Comment by njuettner** [comment]
Thank you Marcus, I was totally ignoring Before as an option. This makes it much easier to read. Awesome!

**Comment by AverageMarcus** [comment]
To be honest, until I noticed you were returning in both cases, I'd have done it the same way.

**Pull request merged**